libraries/AP_Avoidance/AP_Avoidance.h

```diff
@@ -229,3 +231,6 @@ float closest_approach_z(const Location &my_loc,
 namespace AP {
     AP_Avoidance *ap_avoidance();
 };
+
```

Comment:
This namespace should still be here, it should just return null and anyone who uses it should null check.

Reply:
that would completely defeat the purpose of the PR as it would mean the linker would have to leave in the code

Reply:
Our pattern has generally been to make the singleton return a reference if the object is always going to be available.

There's precedent for what @magicrub suggests - but in this case a large amount of code (e.g. the uavcan code) would still be compiled in for no good reason (not without link-time-optimisation and probably not even then...)

Pretty sure the `#if` stuff is the right way to go here.

Reply:
Yeah I agree its ugly, but I've travelled this road many times before and there is no good alternative - if you go the null route you have to include a definition for the AP_ADSB class, which then means you have to include the implementation and the whole ball of yarn unravels. For other subsystems I tried including a very minimal class but that just ended in a mess. It might be possible if we had very good abstraction for the base classes and kept the majority of the implementation hidden, but a lot of our code is not structured that way - we try and put as much functionality in the base class as possible to save flash. So I think we are left with the status quo - ugly, but it works. One other option might be to fix waf so that it only builds things that are actually used, then some of the #defines could go away.